<?xml version="1.0" encoding="utf-8"?>
<launch>
  <arg name="extract_images_sort_ros" default="false" />
  <arg name="extract_images_starmap" default="false" />
  <arg name="extract_images_raw" default="false" />
  
  <arg name="bagfile_basename" default="/media/erl/disk1/orcvio/erl_indoor_rosbags/rs_d435i_atkinson_sparse_8" />
  
  <arg name="bagfile" default="$(arg bagfile_basename).bag" />
  <arg name="outdata_dir" default="/media/erl/disk1/orcvio/erl_indoor_rosbags" />
  <arg name="record_rosbag" default="true" />
  <arg name="starmap_debug" default="false" />
  <arg name="input_image_topic" default="/camera/image_raw" />

  <node name="republish" type="republish" pkg="image_transport" output="screen" args="compressed in:=/camera/infra1/image_rect_raw raw out:=/camera/image_raw" />

  <node pkg="rosbag" type="play" name="rosbag"
        args="--clock $(arg bagfile)"
        required="true"
        />
  <include file="$(find darknet_ros)/launch/darknet_ros.launch" >
    <arg name="image" value="$(arg input_image_topic)" />
    <arg name="bounding_boxes" value="/darknet_ros/bounding_boxes" />
    <arg name="launch_prefix" value=""/>
  </include>
  <node pkg="nodelet" type="nodelet" name="orcvio_nodelet_manager"  args="manager" output="screen"/>
  <node pkg="nodelet" type="nodelet" name="sort_ros"
        args="load sort_ros/SortRos orcvio_nodelet_manager" >
    <param name="image_topic" value="$(arg input_image_topic)"/>
    <param name="bbox_topic" value="/darknet_ros/bounding_boxes" />
    <param name="tracked_bbox_topic" value="tracked_bounding_boxes" />
  </node>
  <node pkg="nodelet" type="nodelet" name="starmap"
        args="load starmap/Starmap orcvio_nodelet_manager"
        output="screen"
        unless="$(arg starmap_debug)"
        >
    <param name="image_topic" value="$(arg input_image_topic)" />
    <param name="to_color_img_flag" type="bool" value="true" />
    <!-- <param name="to_color_img_flag" type="bool" value="false" /> -->
    <param name="bbox_topic" value="/sort_ros/tracked_bounding_boxes" />
    <param name="keypoint_topic" value="/starmap/keypoints" />
    <param name="starmap_model_path" value="$(find starmap)/models/model_cpu-jit.pth" />
    <param name="gpu_id" value="0" type="int" />
    <param name="draw_labels" value="false" type="bool" />
  </node>
  <include file="$(find starmap)/launch/starmap.launch" if="$(arg starmap_debug)" />
  <node pkg="image_view" type="extract_images" name="extract_images_sort_ros"
        if="$(arg extract_images_sort_ros)">
    <remap from="image" to="/sort_ros/detection_image" />
    <param name="filename_format" value="$(arg outdata_dir)/extract_images_sort_ros_%04d.jpg" />
    <param name="sec_per_frame" value="0.1"  />
  </node>
  <node pkg="image_view" type="extract_images" name="extract_images_starmap"
        if="$(arg extract_images_starmap)" >
    <remap from="image" to="/starmap/visualization" />
    <param name="filename_format" value="$(arg outdata_dir)/extract_images_starmap_%04d.jpg" />
    <param name="sec_per_frame" value="0.1"  />
  </node>
  <node if="$(arg record_rosbag)" pkg="rosbag" type="record" name="bag_record"
    args=" -O $(arg outdata_dir)/with_keypoints.bag
              /camera/infra1/camera_info
              /camera/image_raw
              /camera/imu
              /darknet_ros/bounding_boxes
              /sort_ros/tracked_bounding_boxes
              /sort_ros/detection_image
              /starmap/keypoints
              /starmap/visualization
              " />

  <!-- open rviz  -->
  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find starmap)/launch/orcvio_starmap.rviz" />

</launch>
